### PR TITLE
first challenge with quick_sort function

### LINKED ISSUE #8: Challenge: quick_sort
Build an algorithm for sorting sequences.

 Requirements:
quick_sorting algorithm with clear documentation

### PR DESCRIPTION
**Pull request by JumongEric**
QuickSort is a popular and efficient sorting algorithm that follows the divide-and-conquer paradigm. The basic idea behind QuickSort is to select a "pivot" element from the array and partition the other elements into two sub-arrays according to whether they are less than or greater than the pivot. The process is then applied recursively to the sub-arrays.

### CODE REVIEW

**Comment by tvsirius** [comment]
You should not have deleted the check list that was here

I will put it back

**Comment by tvsirius** [comment]
- [x] PR's title includes the challenge's name and language
- [x] PR has short and clear description of the challenge
- [x] PR has appropriate labels and milestones for easy identification
- [x] PR it is assigned to the owner
- [x] reviewers are assigned
- [x] the PR contributes only one focused change
- [x] the branch is up to date with `main`/`master`
- [x] the code works when pulled and run locally
- [x] all conflicts are resolved (if any)
- [x] It is linked to an issue in the appropriate column of the project board
- [x] feedback is addressed (if any, and if it is appropriate feedback.)

## README Documentation

- [x] The solution is documented in `/src/README.md`
- [x] The markdown source is formatted
- [x] Spelling and grammar is correct in all text
- [x] The markdown looks correct when you preview the file
- [x] All links and images work
- [x] The README documents the solution's behavior, strategy(ies) and implementation(s)
- [x] There are use cases in code blocks to illustrate the function's behavior

## Python Files

- [ ] There is a module header
- [ ] There is a module docstring
- [x] File names are in snake_case
- [x] Test files are named `test_<module_name>.py`

## Function Docstring

- [x] Behavior description
- [x] Parameter description
- [x] Return value description
- [x] Include any assumptions
- [x] Include 3 or more (passing!) doctests
- [ ] Include 1-2 use cases (if necessary)

## Function Implementation

- [x] The solution is not is not copied from [djeada](https://github.com/djeada/Algorithms-And-Data-Structures)
- [x] The code is auto-formatted
- [x] The code has no (reasonable) linting mistakes
- [x] Variables are named with snake_case
- [x] The function has a clear and helpful name
- [x] The file's name matches the function name
- [x] The code follows the strategy as simply as possible
- [x] Variable names are clear and helpful
- [x] Comments explain the strategy (if necessary)
- [x] There are type annotations
- [ ] (challenge) The code includes defensive assertions

## Unit Test Suite

- [x] The test class has a helpful name in PascalCase
- [ ] The test class has a docstring
- Each unit test has
  - [x] A helpful name
  - [ ] A clear docstring
  - [x] Only one assertion
- [x] All tests pass
- [ ] There are tests for defensive assertions
- [ ] There are tests for boundary cases
- [ ] Test suite includes black-box and glass-box tests

**Comment by tvsirius** [comment]
add README.md
add working test_quick_sort
when i pull your branch i have to fix it, 
with
```
from quick_sort import quick_sort
```

**Comment by tvsirius** [comment]
Code is good and working, but it could be made better

You are missing:
module and test docstrings
type annotaions
comments marking zero case and it would be better to add coments on logic
use auto formatter (it's easy, just install autoPEP8 and press Shift-Alt-F in VSCode)

File: src/sorting/python/quick_sort/src/quick_sort.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        quick_sort([3,2,5,7])
+        return=[2,3,5,7]
+        quick_sort(("back"))
+        return 'a','b', 'c', 'k' """
+
+    try:
+        if isinstance(iterable, str):
+            sequence = list(iterable)
+        else:
+            sequence= list(iterable)
```
Comment (tvsirius) [inline src/sorting/python/quick_sort/src/quick_sort.py]: WHY?! You will run sequence= list(iterable) in any case, why you use if-else

and it's not common to use try here, we use like
```
assert isinstance(iterable, (str,list,tuple,set)), "Input must be convertible to a list"
```

File: src/sorting/python/quick_sort/src/quick_sort.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
+def quick_sort(iterable):
```
Comment (tvsirius) [inline src/sorting/python/quick_sort/src/quick_sort.py]: type annotation please

Comment (tvsirius) [inline src/sorting/python/quick_sort/src/quick_sort.py]: and a module docstring

File: src/sorting/python/quick_sort/src/quick_sort.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    
+    Parameters:
+        iterable: Iterable object to be sorted.
+        
+    Returns:
+        list: Sorted list.
+        quick_sort([3,2,5,7])
+        return=[2,3,5,7]
+        quick_sort(("back"))
+        return 'a','b', 'c', 'k' """
```
Comment (tvsirius) [inline src/sorting/python/quick_sort/src/quick_sort.py]: there is a format how to write doctest:
```
>>>quick_sort([3,2,5,7])
[2, 3, 5, 7]

>>>quick_sort(("back"))
['a', 'b', 'c', 'k']

File: src/sorting/python/quick_sort/src/quick_sort.py [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+        lower_items=[]
+        equal_items=[]
+    for i in sequence:
+        if i < pivot:
+            lower_items.append(i)
+        elif i== pivot:
+            equal_items.append(i)
+        else:
+            upper_items.append(i)
+    return quick_sort(lower_items) + equal_items + quick_sort(upper_items)
```
Comment (tvsirius) [inline src/sorting/python/quick_sort/src/quick_sort.py:45]: looks like good, maybe some comments will add perfection

**Comment by tvsirius** [comment]
Good, just fix the other issues

**Review by tvsirius — CHANGES_REQUESTED** [review]
Hey Eric, you solve most of my comments, but some left, that requires fixing
Please follow my comments

File: src/sorting/python/quick_sort/src/quick_sort.py (hunk 1 above)
Comment (JumongEric) [reply src/sorting/python/quick_sort/src/quick_sort.py]: Thanks, I appreciate your insight and help

File: src/sorting/python/quick_sort/src/quick_sort.py [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+    quick_sort.py, This module provides a function to perform QuickSort on an iterable.
+    Usage:
+    from quick_sort import quick_sort
+        
+    Returns:List[Union[int, str]]: The sorted listiterable
+        >>>quick_sort([3,2,5,7])
+        [2,3,5,7]
+        >>>quick_sort(("back"))
+        ['a','b', 'c', 'k'] """
```
Comment (tvsirius) [inline src/sorting/python/quick_sort/src/quick_sort.py:19]: look up how this is done in other examples.
you should not use such indent

**Review by tvsirius — APPROVED** [review]
Thats good, we can approve it to move on

Just no docsting and header for moudule (notes for future)
No docstrigns in test (but we speak with other team memebers and also find ok to not add them(
And look up my comment on doctest

Everything else is really good!
